### ISSUE TITLE
Enforce which link types can be used

### ISSUE THREAD

**Issue by xFrednet** [rust-lang/rust-clippy#10176]
### What it does

Rust uses markdown for doc comments. Markdown is a pretty loose standard and behaves differently based on the implementation. Rustdoc supports three different ways to specify a link with a link location:

```
# 1
[<text 1>](https://github.com/rust-lang/rust-clippy)

# 2
[<text 2>][https://github.com/rust-lang/rust-clippy]

# 3
[<text 3>]

[<text 3>]: https://github.com/rust-lang/rust-clippy
```

See: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html

---

This lint should have a configuration, that can specify which links are allowed in doc comments. The configuration could be called: `allowed-doc-link-styles` or something similar, taking a list of items like:
* `InlineBrackets` for `[][]`
* `InlineParenthesis` for `[]()`
* maybe `Reference` for the third case.

By default, I'd allow `InlineParenthesis` and `Reference` as these are normal Markdown links. 

---

This should be a relatively simple lint, as Clippy already has a markdown parser for doc comments, however, it might require some handling to correctly detect which type of link is actually used

### Lint Name

comment_link_style

### Category

restriction

### Advantage

- Unifies the codebase

### Drawbacks

- Might cause many reports, when it's not a restriction lint

### Example

```rust
/// # 1
/// [<text 1>](`Option`)
/// 
/// # 2
/// [<text 2>][`Option`]
/// 
/// # 3
/// [<text 3>]
/// 
/// [<text 3>]: `Option`
struct CoolItem;
```

Could be written as:

```rust
/// # 1
/// [<text 1>](`Option`)
/// 
/// # 2
/// [<text 2>](`Option`)
/// 
/// # 3
/// [<text 3>]
/// 
/// [<text 3>]: `Option`
```
